### PR TITLE
Config: Race fix (draft)

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

This includes a hotfix for a potential race condition when the engine starts and stops a global service and multiple services are interacting with the configuration system. 

Moves config fields to a subtype to be protected by a mutex and not accessible out of the package. Not idiomatic but added getter and setter methods ad hoc. 

Open as a draft to gauge feedback and to get a better direction. 

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Travis with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: engine/events.go [hunk 1]
```diff
@@ -308,10 +308,10 @@ func EventManger() {
 
 // IsValidExchange validates the exchange
 func IsValidExchange(exchangeName string) bool {
-	cfg := config.GetConfig()
-	for x := range cfg.Exchanges {
-		if strings.EqualFold(cfg.Exchanges[x].Name, exchangeName) && cfg.Exchanges[x].Enabled {
-			return true
+	exchanges := config.GetConfig().GetAllExchangeConfigs()
```
Comment (Rots) [inline engine/events.go:311]: a possibly better approach (to take care of the race problem) is to eliminate all the uses of `config.GetConfig()` and always get it from the *Engine object instead

File: engine/engine.go [hunk 2]
```diff
@@ -52,13 +52,14 @@ var (
 // New starts a new engine
 func New() (*Engine, error) {
 	var b Engine
-	b.Config = &config.Cfg
+	b.Config = config.GetConfig()
```
Comment (Rots) [inline engine/engine.go:55]: I think the purpose here is to always have a clean config to start with... the GetConfig() is risky as the returned config may already be used. So I think the previous was correct.
If you want to make the construction explicit, you can add a `config.New()` function and use that instead.

**Review by Rots — COMMENTED** [review]
Were you able to reproduce the race condition with a test?
I'm not convinced this improves the code...

**Comment by thrasher-** [comment]
Closed due to inactivity and fixes implemented in the backtester PR

**Pull request closed without merging**